src/Microsoft.Identity.Client/MsalException.cs

```diff
@@ -1,122 +1,188 @@
+        /// exception handling. Values for this code are typically provided in constant strings in the derived exceptions types
+        /// with explanations of mitigation.
+        /// </summary>
+        public string ErrorCode { get; private set; }
+
+        /// <summary>
+        /// Creates and returns a string representation of the current exception.
+        /// </summary>
+        /// <returns>A string representation of the current exception.</returns>
+        public override string ToString()
+        {
+            return base.ToString() + string.Format(CultureInfo.InvariantCulture, "\n\tErrorCode: {0}", ErrorCode);
+        }
+
+        #region SERIALIZATION
+
+        private const string ExceptionTypeKey = "type";
+        private const string ErrorCodeKey = "error_code";
+        private const string ErrorDescriptionKey = "error_description";
+
+        internal virtual void PopulateJson(JObject jobj)
+        {
+            jobj[ExceptionTypeKey] = GetType().Name;
+        }
+
+        internal virtual void PopulateObjectFromJson(JObject jobj)
+        {
+        }
+
+        /// <summary>
+        /// Allows serialization of most values of the exception into JSON.
+        /// </summary>
+        /// <returns></returns>
+        public string ToJsonString()
+        {
+            JObject jobj = new JObject();
+            PopulateJson(jobj);
+            return jobj.ToString();
+        }
+
+        private static readonly Lazy<Dictionary<string, Type>> s_typeNameToType = new Lazy<Dictionary<string, Type>>(() =>
+        {
+            return new Dictionary<string, Type>
+            {
+                { typeof(MsalException).Name, typeof(MsalException) },
+                { typeof(MsalClientException).Name, typeof(MsalClientException) },
+                { typeof(MsalServiceException).Name, typeof(MsalServiceException) },
+                { typeof(MsalUiRequiredException).Name, typeof(MsalUiRequiredException) },
+            };
+        });
+
+        /// <summary>
+        /// Allows re-hydration of the MsalException (or one of its derived types) from JSON generated by ToJsonString().
+        /// </summary>
+        /// <param name="json"></param>
+        /// <returns></returns>
+        public static MsalException FromJsonString(string json)
+        {
+            JObject jobj = JObject.Parse(json);
+            string type = jobj.Value<string>(ExceptionTypeKey);
+
+            if (s_typeNameToType.Value.TryGetValue(type, out Type exceptionType))
+            {
+                string errorCode = JsonUtils.GetExistingOrEmptyString(jobj, ErrorCodeKey);
+                string errorMessage = JsonUtils.GetExistingOrEmptyString(jobj, ErrorDescriptionKey);
+
+                MsalException ex = Activator.CreateInstance(exceptionType, errorCode, errorMessage) as MsalException;
+                ex.PopulateObjectFromJson(jobj);
+                return ex;
+            }
+
+            throw new MsalException();
```

Comment:
>throw new [](start = 12, length = 9)

shouldn't this return ? Alternatively the method should be ThrowExceptionFromJsonString()

Reply:
This is the case where we tried to deserialize a json that has a type in it we don't understand.  so we should definitely throw an exception here, shouldn't we?  thus, the todo below about adding in a real error message (i was waiting on the core error message merge to come in before i added the message properly)